generator/sponsored_dataset_inl.hpp

```diff
@@ -0,0 +1,179 @@
+{
+  LoadData(dataSource, addressReferencePath);
+}
+
+template <typename SponsoredObject>
+typename SponsoredDataset<SponsoredObject>::Object const &
+SponsoredDataset<SponsoredObject>::GetObjectById(ObjectId id) const
+{
+  auto const it = m_objects.find(id);
+  CHECK(it != end(m_objects), ("Got wrong object id:", id));
+  return it->second;
+}
+
+template <typename SponsoredObject>
+typename SponsoredDataset<SponsoredObject>::Object &
+SponsoredDataset<SponsoredObject>::GetObjectById(ObjectId id)
+{
+  auto const it = m_objects.find(id);
+  CHECK(it != end(m_objects), ("Got wrong object id:", id));
+  return it->second;
+}
+
+template <typename SponsoredObject>
+void SponsoredDataset<SponsoredObject>::BuildOsmObjects(function<void(FeatureBuilder1 &)> const & fn) const
+{
+  for (auto const & item : m_objects)
+    BuildObject(item.second, fn);
+}
+
+template <typename SponsoredObject>
+typename SponsoredDataset<SponsoredObject>::ObjectId
+SponsoredDataset<SponsoredObject>::FindMatchingObjectId(FeatureBuilder1 const & fb) const
+{
+  if (NecessaryMatchingConditionHolds(fb))
+    return FindMatchingObjectIdImpl(fb);
+  return Object::InvalidObjectId();
+}
+
+template <typename SponsoredObject>
+vector<typename SponsoredDataset<SponsoredObject>::ObjectId>
+SponsoredDataset<SponsoredObject>::GetNearestObjects(ms::LatLon const & latLon, size_t const limit,
+                                                     double const maxDistanceMeters /* = 0.0 */) const
+{
+  namespace bgi = boost::geometry::index;
+
+  vector<ObjectId> indexes;
+  for_each(bgi::qbegin(m_rtree, bgi::nearest(Point(latLon.lat, latLon.lon), limit)),
+           bgi::qend(m_rtree), [this, &latLon, &indexes, maxDistanceMeters](Value const & v)
+           {
+             auto const & object = GetObjectById(v.second);
+             double const dist = ms::DistanceOnEarth(latLon, object.m_latLon);
+             if (maxDistanceMeters != 0.0 && dist > maxDistanceMeters /* max distance in meters */)
+               return;
+
+             indexes.emplace_back(v.second);
+           });
+
+  return indexes;
+}
+
+template <typename SponsoredObject>
+void SponsoredDataset<SponsoredObject>::LoadData(istream & src, string const & addressReferencePath)
+{
+  m_objects.clear();
+  m_rtree.clear();
+
+  for (string line; getline(src, line);)
+  {
+    Object hotel(line);
+    m_objects.emplace(hotel.m_id, hotel);
+  }
+
+  // Try to get object address from existing MWMs.
+  if (!addressReferencePath.empty())
+  {
+    LOG(LINFO, ("Reference addresses for sponsored objects", addressReferencePath));
+    Platform & platform = GetPlatform();
+    string const backupPath = platform.WritableDir();
+
+    // MWMs can be loaded only from a writebledir or from a resourcedir,
+    // changig resourcedir can lead to probles with classificator, so
+    // we change writebledir.
+    platform.SetWritableDirForTests(addressReferencePath);
```

Comment:
Well, yes. I've investigated this so the comment appeared. I think we can rename it (remove ForTests part).